src/schedule.py

```diff
@@ -86,20 +88,97 @@ def cfg_to_dfg(cfg, graphs, latency):
 
     return cfg_node_to_dfg_map
 
+def assign_time_of_execution(graph):
+    """
+    Calculates when each operation will take place,
+    overwriting the "dist" attribute from before.
+    """
+    for node in graph.nodes:
+        graph.nodes[node]["dist"] =  0
+    graph = nx.reverse(graph)
+    for gen in list(nx.topological_generations(graph))[1:]:
+        max_weight = 0
+        for node in gen:
+            in_edge = list(graph.in_edges(node))[0]
+            max_weight = max(max_weight, graph.edges[in_edge]["weight"])
+        for node in gen:
+            in_edge = list(graph.in_edges(node))[0]
+            graph.edges[in_edge]["weight"] = max_weight
+    max_dist = 0
+    end_node = list(nx.topological_generations(graph))[0][0]
+    for node in graph.nodes:
+        graph.nodes[node]["dist"] = nx.dijkstra_path_length(graph, end_node, node)
+        max_dist = max(max_dist, graph.nodes[node]["dist"])
+    for node in graph.nodes:
+        # mirroring operation
+        graph.nodes[node]["dist"] = (graph.nodes[node]["dist"] - max_dist) * -1
+    graph = nx.reverse(graph)
+    return graph, max_dist
+
 def assign_upstream_path_lengths(graph):
     """
     Assigns the longest path to each node in the graph.
-    Currently ignores actual latencies of nodes.
-    TODO: change this to dijkstra
+    Uses Dijkstra to take operation latency into account.
+    Params:
+    graph: nx.DiGraph
+        The computation graph
     """
-    for node in graph:
+    for node in graph.nodes:
         graph.nodes[node]["dist"] =  0
-    for i, generations in enumerate(nx.topological_generations(graph)):
-        for node in generations:
-            graph.nodes[node]["dist"] = max(i, graph.nodes[node]["dist"])
-    
+    q = deque()
+    for node in list(nx.topological_generations(graph))[0]:
```

Comment:
Ok so this is LIke ASAP, where we define distancne as longest distance from any of the start nodes.

Reply:
yes, basically just walk down the tree and take max to keep updating distance.